examples/fluids/src/matops.c

```diff
@@ -89,55 +88,60 @@ PetscErrorCode MatGetDiag_Ceed(Mat A, Vec D) {
   CeedVectorTakeArray(op_apply_ctx->y_ceed, MemTypeP2C(mem_type), NULL);
   PetscCall(VecRestoreArrayAndMemType(Y_loc, &y));
   PetscCall(VecZeroEntries(D));
-  PetscCall(DMLocalToGlobal(op_apply_ctx->dm, Y_loc, ADD_VALUES, D));
+  PetscCall(DMLocalToGlobal(op_apply_ctx->dm_y, Y_loc, ADD_VALUES, D));
 
-  if (!op_apply_ctx->Y_loc) PetscCall(DMRestoreLocalVector(op_apply_ctx->dm, &Y_loc));
+  if (!op_apply_ctx->Y_loc) PetscCall(DMRestoreLocalVector(op_apply_ctx->dm_y, &Y_loc));
   PetscFunctionReturn(0);
 };
 
 // -----------------------------------------------------------------------------
-// This function uses libCEED to compute the action of the Laplacian with Dirichlet boundary conditions
+// Applies CeedOperator to input and output Vecs
+// If either X or Y are NULL, assume that op_apply_ctx->{x,y}_ceed hold their own data
 // -----------------------------------------------------------------------------
 PetscErrorCode ApplyLocal_Ceed(Vec X, Vec Y, MatopApplyContext op_apply_ctx) {
```

Comment:
Also, a question of whether I should make accommodation for `X` or `Y` being a local vector with `GlobalToLocal`/`LocalToGlobal` having already been applied. That comes up with the diffusive flux projection, and the strain rate and vorticity projection; in both cases the solution state vector in `IFunction_NS` needs to be localized for the residual operator and is thus already available, so doing the `GlobalToLocal` again is wasteful (and probably expensive).

Obviously this would further complicate the function, which I'm fine with, but maybe don't see potential future issues with.

Reply:
As in have a different function for every combination of inputs, and call which one we need at that time? That'd require an outer product of the options. Including X or Y being localized, that'd be 9 total. 

Not sure if the concern is performance or style/robustness. If it's performance, we write a generic `ApplyLocal_Ceed` with boolean inputs for which input/output combination it is, then write individual smaller functions calling the generic function with constant arguments which would hopefully inline and optimize well. Similar to what we do for QFunctions with different state variables.

Reply:
> Can we nest these function calls so we just call the one at the correct level instead of having a lot of logic to figure out what we've been handed?

Ok, I think I understand what you actually meant.

Generically, inputs only need to go through:
```mermaid
flowchart LR
    VG1[VecGlobal] 
    VL1[VecLocal]
    VL2[VecLocal] 
    CV[CeedVector]
    VG1 -->|GlobalToLocal| VL1 -->|SetArray| CV -->|TakeArray| VL2 
```

Outputs only need to go through: 
```mermaid
flowchart LR
    VL1[VecLocal]
    VG2[VecGlobal] 
    VL2[VecLocal] 
    CV[CeedVector]
    VL1 -->|SetArray| CV -->|TakeArray| VL2 -->|LocalToGlobal| VG2
```

So we could have a subfunction on the overlap like this:
```mermaid
flowchart LR
    IVG1[InVecGlobal] 
    OVG2[OutVecGlobal] 

    subgraph one
    IVL1[InVecLocal]
    ICV[ InCeedVector]
    IVL2[InVecLocal] 

    OVL1[OutVecLocal]
    OCV[ OutCeedVector]
    OVL2[OutVecLocal] 
    end

    IVG1 -->|GlobalToLocal| IVL1 -->|SetArray| ICV -->|TakeArray| IVL2
    OVL1 -->|SetArray| OCV -->|TakeArray| OVL2 -->|LocalToGlobal| OVG2
```
The box represents a sub function, taking the local input/output vectors as arguments. The entire diagram would be the equivalent of the current `ApplyLocal_Ceed`. We'd still need internal logic in the function to handle pre-set CeedVectors.

That code would look something like this:

<details>
<summary>Details</summary>

```c
PetscErrorCode ApplyLocal_Ceed_VecLocal(Vec X_loc, Vec Y_loc, MatopApplyContext op_apply_ctx) {
  PetscScalar *x, *y;
  PetscMemType x_mem_type, y_mem_type;

  if (X_loc) {
    PetscCall(VecGetArrayReadAndMemType(X_loc, (const PetscScalar **)&x, &x_mem_type));
    CeedVectorSetArray(op_apply_ctx->x_ceed, MemTypeP2C(x_mem_type), CEED_USE_POINTER, x);
  }

  if (Y_loc) {
    PetscCall(VecGetArrayAndMemType(Y_loc, &y, &y_mem_type));
    CeedVectorSetArray(op_apply_ctx->y_ceed, MemTypeP2C(y_mem_type), CEED_USE_POINTER, y);
  }

  CeedOperatorApply(op_apply_ctx->op, op_apply_ctx->x_ceed, op_apply_ctx->y_ceed, CEED_REQUEST_IMMEDIATE);

  if (X_loc) {
    CeedVectorTakeArray(op_apply_ctx->x_ceed, MemTypeP2C(x_mem_type), NULL);
    PetscCall(VecRestoreArrayReadAndMemType(X_loc, (const PetscScalar **)&x));
  }

  if (Y_loc) {
    CeedVectorTakeArray(op_apply_ctx->y_ceed, MemTypeP2C(y_mem_type), NULL);
    PetscCall(VecRestoreArrayAndMemType(Y_loc, &y));
  }
}

PetscErrorCode ApplyLocal_CeedVecGlobal(Vec X, Vec Y, MatopApplyContext op_apply_ctx) {
  Vec          Y_loc = NULL, X_loc = NULL;
  PetscFunctionBeginUser;

  // Setup operator input
  if (op_apply_ctx->X_loc) X_loc = op_apply_ctx->X_loc;
  if (X) {
    if (!X_loc) PetscCall(DMGetLocalVector(op_apply_ctx->dm_x, &X_loc));

    PetscCall(DMGlobalToLocal(op_apply_ctx->dm_x, X, INSERT_VALUES, X_loc));
  }

  // Setup operator output
  if (op_apply_ctx->Y_loc) Y_loc = op_apply_ctx->Y_loc;
  if (Y) {
    if (!Y_loc) PetscCall(DMGetLocalVector(op_apply_ctx->dm_x, &Y_loc));
  }

  PetscCall(ApplyLocal_Ceed_VecLocal(X_loc, Y_loc, op_apply_ctx));

  if (!op_apply_ctx->X_loc) PetscCall(DMRestoreLocalVector(op_apply_ctx->dm_x, &X_loc));

  if (Y) {  // Process operator output
    PetscCall(VecZeroEntries(Y));
    PetscCall(DMLocalToGlobal(op_apply_ctx->dm_y, Y_loc, ADD_VALUES, Y));
  }
  if (!op_apply_ctx->Y_loc) PetscCall(DMRestoreLocalVector(op_apply_ctx->dm_y, &Y_loc));
  PetscFunctionReturn(0);
};

```

</details>

However, I don't know how useful this is, as calling the nested functions directly doesn't really seem feasible. Much of the time, the input and output flow are not symmetric. 

For example, the statistics processing takes in a pre-set CeedVector, so the entire input flow needs to be skipped, but the entire output flow is needed. To directly call a function, it'd need to have a function signature `ApplyLocal_CV_VG(CeedVector x_ceed, Vec Y)`, but I don't see a way of cleanly implementing something like that.

Reply:
What about taking explicit arguments for the input and output global, local, and ceed vectors (some of which can be NULL), instead of tucking them into the context? Maybe it's a non-issue, but I'd like to avoid mutating the context in order to call the function that applies the operator, and I don't love using heuristics to figure out what is being passed (better to use arguments that mean exactly one thing).

Reply:
I guess my idea, which might be overly ambitious, is to define the input and output of the operation during setup, so that when it's called you give it the desired input and output object (be it a Global Vec, Local Vec, or CeedVector) and it does the operation.

So maybe a solution that addresses your concerns would be something like this:
 - At Context creation, we 
    - get the input and output global, local, and ceed vectors, which may be NULL
    - explicitly ask what the type of the active input and output object is
 - At apply time, we take in void pointers and convert it to a ceed vector via casting and helper functions
 
This would eliminate any need for the user to mutate the context objects directly, avoid heuristics by explicitly storing what should be coming in as input and output, and still be flexible enough to apply to basically any use case we have.

```c
typedef enum  {
  MATOPVEC_GLOBAL = 1,
  MATOPVEC_LOCAL  = 2,
  MATOPVEC_CEED   = 3,
} MatopActiveVectorType;

typedef struct OperatorApplyContext2_ *MatopApplyContext2;
struct OperatorApplyContext2_ {
  DM                    dm_x, dm_y;
  Vec                   X_loc, Y_loc;
  CeedVector            x_ceed, y_ceed;
  MatopActiveVectorType in_type, out_type;
  CeedOperator          op;
  Ceed                  ceed;
};

PetscErrorCode InputCeedVectorFromVecGlobal(CeedVector x_ceed, Vec X, MatopApplyContext2 op_apply_ctx);
PetscErrorCode OutputCeedVectorFromVecGlobal(CeedVector y_ceed, Vec Y, MatopApplyContext2 op_apply_ctx);
PetscErrorCode CeedVectorFromVecLocal(CeedVector x_ceed, Vec X_loc);

PetscErrorCode InputVecGlobalFromCeedVector(Vec X, CeedVector x_ceed, MatopApplyContext2 op_apply_ctx);
PetscErrorCode OutputVecGlobalFromCeedVector(Vec Y, CeedVector y_ceed, MatopApplyContext2 op_apply_ctx);
PetscErrorCode VecLocalFromCeedVector(Vec X_loc, CeedVector x_ceed);

PetscErrorCode ApplyLocal_Ceed2(void *in, void *out, MatopApplyContext2 op_apply_ctx) {
  CeedVector   x_ceed = op_apply_ctx->x_ceed, y_ceed = op_apply_ctx->y_ceed;
  PetscFunctionBeginUser;

  switch (op_apply_ctx->in_type) {
    case MATOPVEC_GLOBAL:
      InputCeedVectorFromVecGlobal(x_ceed, (Vec)in, op_apply_ctx);
    break;
    case MATOPVEC_LOCAL:
      CeedVectorFromVecLocal(x_ceed, (Vec)in);
    break;
    case MATOPVEC_CEED:
      x_ceed = (CeedVector)in;
    break;
  }

  switch (op_apply_ctx->out_type) {
    case MATOPVEC_GLOBAL:
      OutputCeedVectorFromVecGlobal(y_ceed, (Vec)out, op_apply_ctx);
    break;
    case MATOPVEC_LOCAL:
      CeedVectorFromVecLocal(y_ceed, (Vec)out);
    break;
    case MATOPVEC_CEED:
      y_ceed = (CeedVector)out;
    break;
  }

  // Apply libCEED operator
  CeedOperatorApply(op_apply_ctx->op, x_ceed, y_ceed, CEED_REQUEST_IMMEDIATE);

  switch (op_apply_ctx->in_type) {
    case MATOPVEC_GLOBAL:
      InputVecGlobalFromCeedVector((Vec)in, x_ceed, op_apply_ctx);
    break;
    case MATOPVEC_LOCAL:
      VecLocalFromCeedVector((Vec)in, x_ceed);
    break;
    case MATOPVEC_CEED:
    break;
  }

  switch (op_apply_ctx->out_type) {
    case MATOPVEC_GLOBAL:
      OutputVecGlobalFromCeedVector((Vec)out, y_ceed, op_apply_ctx);
    break;
    case MATOPVEC_LOCAL:
      VecLocalFromCeedVector((Vec)out, y_ceed);
    break;
    case MATOPVEC_CEED:
    break;
  }
  PetscFunctionReturn(0);
};
```

Reply:
This sounds a heck of a lot like what I built for Ratel? I handled this by having many separate contexts and optionally scattering/gathering based upon if a global vector was passed.

Reply:
Can you point me at where you're doing that? Quick grep of `GlobalToLocal` doesn't show any conditionals around it that I can tell.

Reply:
Right now the condition is only on `LocalToGlobal` as we always have an input global vector, but the same logic could be used for inputs or transfering arrays from PETSc to libCEED vecs

Reply:
Ah, ok, now I can see it ([example](https://gitlab.com/micromorph/ratel/-/blob/main/src/ratel-diagnostic.c#L210-213)). So you're directly accessing the context member here.

I guess the primary difference between what's done in Ratel and what I'm generally proposing is just more flexibility in the inputs and outputs. Fluids "requires" more flexibility, as currently the origin for CeedOperator inputs range the entire hierarchy (Local Vec, Global Vec, and CeedVector) and the outputs can be CeedVector or a Global Vec. 

Additionally, I added in the option for Local vectors to be Get/Restored rather than requiring that they be given to the context. I believe that allows the code to be more memory efficient (don't get memory until we need it), but I admittedly haven't fully looked into that.

Reply:
Right, in Ratel the output can be a local Vec or a global Vec. Our inputs include a global Vec (though that can be easily optional) and CeedVectors managed by the operator.

I think the big difference here is that rather than trying to be generic on Vec vs CeedVector, I went the route in Ratel of keeping all data in a Vec of some sort and then being generic or if the Vec was global or local. I see the job of Ratel as wrapping all the libCEED bits into appropriate PETSc types.

In Ratel I think we are only using the memory we need as well. The same local Vec is usually referenced across many contexts. This lets us minimize the expense of setting boundary values and directly use as few local Vecs as possible.

Long term I care what direction this generalized interface goes because I want to add fluids and fluid-solid interactions into Ratel, so the less dissimilar the two codes are, the better.

Reply:
I'd like to second that last point. I think the exact interface here is prone to bikeshedding and we should strive to make it easier to merge fluids into Ratel so we can do FSI with good solvers (which I think would be a unique capability) in the medium term. Fluids has/will have a lot more auxiliary operators than (non-multiscale) solids: distance function, stats collection and projection, diffusive flux, SGS, differential filter. I think it's okay if fluids gets out ahead so long as it's something Ratel can readily adopt. I hope we can upstream these utilities to PETSc. I think it'll make our lives a lot easier and make the tools more accessible.

Reply:
Yeah, I don't want to fall into bikesheading for sure. I'm trying to understand the requirements so I can push the Ratel and fluids codes in the same direction. I've been trying to pull features from fluids to Ratel to smooth the future work of putting fluids in Ratel. 

I think my only real objection here is just trying to avoid using void in a function signature if we can. Vecs would let us avoid that I think?

Reply:
> I'm trying to understand the requirements

It depends on what you mean by requirements. The fluids code has worked without these kinds of wrapped CeedOperatorApply functions, and absolutely can continue to. If we want to be *able* to replace every instance of `CeedOperatorApply` with a wrapped function, then it needs to allow for input and output to be defined everywhere on the hierarchy (Global Vec, Local Vec, and CeedVector). Keyword there is "able", as I don't know if we actually want to do that, but I'd lean in that direction unless it's just CeedVectors for both input and output.

Other "wants" that have been mentioned:
- avoid mutating the context in order to call the function that applies the operator
- avoid using heuristics to figure out what is being passed (better to use arguments that mean exactly one thing)
- avoid `void` function parameters (I'm guessing for type checking reasons?)
- make fluids easier to merge into Ratel

#### Using [my latest proposal](https://github.com/CEED/libCEED/pull/1147#discussion_r1087291726) as a jumping off point:
I think to avoid heuristics in this case, setting an enum at context creation to explicitly declare what's going to be used as input and output works well. That frees up `{X,Y}_loc` and `{x,y}_ceed` to be NULL depending on the situation. But that ends up with more divergence from Ratel's code (which the current `matops.c` is heavily based on /copied from).

Using Vec function parameters would allow us to avoid `void` function parameters, but then the question is what to do about CeedVector input/output. We could just allow those to be directly manipulated via the context, but then the interface is inconsistent. I honestly don't know which of the two is better or worse, or if there's another option.

#### Alternative to my latest proposal:
If maximum priority is making sure Ratel and the fluids example stay close to each other for future merging (and we don't want to make changes to Ratel's API), then I think we have to ditch CeedVector input/output from being a part of the interface; it'll just have to be manually coded up like it has been (which isn't that bad). We'd also need to ditch using Get/Restore for Local Vecs (also not too bad). Then we'd match Ratel's behavior as far as handling Global Vec and Local Vec as inputs and outputs.

Reply:
A point of note - it would be pretty easy for me to modify Ratel to get and release l-vecs when the cached l-vec isn't important. 

Most output l-vecs and input l-vecs where there is no Dirichlet conditions would be the two main cases where that would occur.

Reply:
See: https://gitlab.com/micromorph/ratel/-/merge_requests/329

Reply:
Thanks for writing this out. I don't think there's harm in always using `Vec` as owner. I also think it would be fine to take arguments `Vec X, CeedVector x_ceed, CeedVector y_ceed, Vec Y` and allow `NULL` to be used for some of these arguments. That is, if `X` and `Y` are NULL, then you're just applying the operator on the `CeedVector`s. If everything is non-NULL, the work-space is being provided, but the operation maps from `X` to `Y`.

Reply:
I didn't add the option to exclude `X` to Ratel since we don't currently need it, but here is how I'd do that modification given our current code:
```C
PetscErrorCode RatelApplyAddLocalCeedOp(Vec X, Vec Y, RatelOperatorApplyContext ctx) {
  Ratel        ratel = ctx->ratel;
  Vec          X_loc = ctx->X_loc, Y_loc = ctx->Y_loc;
  PetscMemType x_mem_type, y_mem_type;

  PetscFunctionBeginUser;

  // Vec validity check
  if ((!X && !ctx->X_loc) || (!Y && !ctx->Y_loc)) {
    // LCOV_EXCL_START
    SETERRQ(ratel->comm, PETSC_ERR_SUP, "A local or global Vec must be provided for the input and output");
    // LCOV_EXCL_STOP
  }

  // Get local vectors, as needed
  if (!ctx->X_loc) PetscCall(DMGetLocalVector(ctx->dm_x, &X_loc));
  if (!ctx->Y_loc) PetscCall(DMGetLocalVector(ctx->dm_y, &Y_loc));

  // Global-to-local
  if (X) PetscCall(DMGlobalToLocal(ctx->dm_x, X, INSERT_VALUES, X_loc));

  // Setup libCEED vectors
  PetscCall(RatelVecReadP2C(ratel, X_loc, &x_mem_type, ctx->x_loc));
  PetscCall(RatelVecP2C(ratel, Y_loc, &y_mem_type, ctx->y_loc));

  // Apply libCEED operator
  RatelCeedCall(ratel, CeedOperatorApply(ctx->op, ctx->x_loc, ctx->y_loc, CEED_REQUEST_IMMEDIATE));

  // Restore PETSc vectors
  PetscCall(RatelVecReadC2P(ratel, ctx->x_loc, x_mem_type, X_loc));
  PetscCall(RatelVecC2P(ratel, ctx->y_loc, y_mem_type, Y_loc));

  // Local-to-global
  if (Y) PetscCall(DMLocalToGlobal(ctx->dm_y, Y_loc, ADD_VALUES, Y));

  // Restore local vectors, as needed
  if (!ctx->X_loc) PetscCall(DMRestoreLocalVector(ctx->dm_x, &X_loc));
  if (!ctx->Y_loc) PetscCall(DMRestoreLocalVector(ctx->dm_y, &Y_loc));

  PetscFunctionReturn(0);
}
```
I think that looks pretty close in function/form to the current code in this PR

Reply:
> Thanks for writing this out. I don't think there's harm in always using `Vec` as owner. I also think it would be fine to take arguments `Vec X, CeedVector x_ceed, CeedVector y_ceed, Vec Y` and allow `NULL` to be used for some of these arguments. That is, if `X` and `Y` are NULL, then you're just applying the operator on the `CeedVector`s. If everything is non-NULL, the work-space is being provided, but the operation maps from `X` to `Y`.

And if `X` and `Y` are non-NULL, but `x_ceed` and `y_ceed` are, we'd use the `CeedVector`s provided in the context?

Reply:
I'm not quite sure how to proceed on this (and is close to holding up #1149). It kinda seems like we have 2.5 objectives:
1) allow for easy migration to Ratel, 
2) generalize the operator application code to different forms of inputs and outputs

2.5) avoid heuristics when possible

and a Pareto front of different solutions:

1) [Explicitly define inputs and outputs at context creation](https://github.com/CEED/libCEED/pull/1147#discussion_r1087291726). That description has the `void *` parameter issue, but we can just create multiple typed-wrappers to address that (ie. `ApplyLocal_Ceed_P2C(Vec, CeedVector, ctx)`).
- Ratel objective: Requires changing both context creation and operator application function calls,
- Generalization objective: Generalizes to every input/output combination  (including possible expansion to not-foreseen ones via expanding the enum)
- Heuristic objective: Explicit about what it's inputs and outputs are

2) Change `ApplyLocal` signature to `(Vec X, CeedVector x_ceed, CeedVector y_ceed, Vec Y, ctx)` (I believe this is what Jed suggested [here](https://github.com/CEED/libCEED/pull/1147#discussion_r1088151458)). 
- Ratel objective: Requires changing operator application function calls,
- Generalization objective: Can generalize to every input/output combination if we use NULL/non-NULL arguments in the context 
- Heuristic objective: Requires heuristics, but the extend is dependent on what input/output combinations we want to support. It would "hide" the use of localized `Vec` objects as input/output by having it in the context.

3) Keeping everything mostly the same, but change the input/output `Vec` to be optional (see Jeremy's code paste [here](https://github.com/CEED/libCEED/pull/1147#discussion_r1088268957)). This is the closest to Ratel, but doesn't allow for CeedVector inputs This 
- Ratel objective: Minor change to operator application function, but signature stays the same
- Generalization objective: Cannot handle CeedVector inputs  or outputs (be they `NULL` or actually having data).
- Heuristic objective: Doesn't explicitly mention it's inputs and outputs, but it's pretty straightforward in heuristics.

Given I think these form a Pareto front, all solution are valid. The choice between then all is just down to how we prioritize the objectives. I'm partial to suggestion 1, but I have a tendency to generalize functions to a fault.

Reply:
I agree all options are *valid*. My two main points here are

1) C is bad enough about type safety. I don't like using `void *` in an interface unless we absolutely must. Wrappers are fine so long as we never expose the non-wrapped version of the function in a header.

2) Long term we want to wrap the libCEED functionality into DMs so the user never needs to touch PETSc objects. I think using PETSc Vecs makes that future transition easier. I'm pushing Ratel this direction to reduce my future pain in this effort.

Reply:
> Wrappers are fine so long as we never expose the non-wrapped version of the function in a header.

Yeah, the intent was to only have the non-void typed wrappers in the API.

> I think using PETSc Vecs makes that future transition easier.

The issue here is that sometimes we work with quadrature point-level data (statistics for example), so Vecs aren't possible/ideal for that.

However, if the future plan is to wrap most things into DMs, then I'll just proceed with option 3. That'll mean that the `CeedVector` input/output applications will have to be coded manually, but they're rare enough. I'll probably bring in equivalents of `RatelVecReadP2C` and friends to make those cases easier/cleaner to implement.

Reply:
Since `PetscFE` is fully discrete (specifies quadrature), we could add a `DMPlexCreateQuadratureVector`.

One more possibility is `RatelApplyAddLocalCeedOp(Vec X, CeedVector X_ceed, Vec Y, CeedVector Y_ceed, RatelOperatorApplyContext ctx)`, in which at most one of `X` and `X_ceed` can be NULL. I prefer this to `void*` because it's still type checked and intent is more explicit.

Reply:
Small change in Ratel for clarity: https://gitlab.com/micromorph/ratel/-/commit/283206d82b406db70993966f9adfff77cab6c05d